fix(*): Use go env GOPATH for install golangci-lint

Meoop opened the pull request:
<!--  Thanks for sending a pull request! See below for tips! -->

**What this PR does / why we need it**:

在新版本的 Golang 中，GOPATH 环境变量可以不设置，当系统未设置 GOPATH 环境变量时，golangci-lint 将安装到 /bin 目录下，安装错位置。使用 `go env GOPATH` 命令获取 GOPATH 路径可以应对所有的场景。

参考 golangci-lint 将 `GOPATH` 换成 `go env GOPATH`： https://github.com/golangci/golangci-lint/pull/319

**Which issue(s) this PR is related to** *(optional, link to 3rd issue(s))*:

Fixes #

Reference to #
<!-- 填在 Fixes，PR 合并就会关 issue。填在 Reference to 会关联 issue，不会联动关闭。-->

**Special notes for your reviewer**:

/cc @bbbmj 

<!-- Please answer the following questions during the code freeze, and delete this line.
**Code freeze questions**

1. What causes this PR to not be merged before code freeze?
2. Why this PR is absolutely necessary for this version? Paste a screenshot of smoke testing docs if you could.
3. What's the effects after merging it?
4. Is there anyway we can skip this to not affect the overall process?
-->

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->

```release-note
NONE
```

<!--  Thanks for sending a pull request! Here are some tips:

1. https://github.com/caicloud/engineering/blob/master/guidelines/review_conventions.md      <-- what is the review process looks like
2. https://github.com/caicloud/engineering/blob/master/guidelines/git_commit_conventions.md  <-- how to structure your git commit
3. https://github.com/caicloud/engineering/blob/master/guidelines/caicloud_bot.md            <-- how to work with caicloud bot

Other tips:

If this is your first contribution, read our Getting Started guide https://github.com/caicloud/engineering/blob/master/guidelines/README.md
-->

Meoop commented:
/cc @bbbmj

supereagle commented on Makefile:
```diff
@@ -70,6 +70,7 @@ CPUS_AVAILABLE ?= $(shell sh hack/read_cpus_available.sh)
 DOCKER_LABELS ?= git-describe="$(shell date -u +v%Y%m%d)-$(shell git describe --tags --always --dirty)"
 
 # Golang standard bin directory.
+GOPATH ?= $(shell go env GOPATH)
```
What about including this in #5?  @Meoop

Meoop replied:
lgtm

supereagle replied:
在流水线执行的时候，代码构建阶段的环境没有装 go，而是通过启 go 容器运行的，会有问题。
之前将 `PKGS` 变量去掉，就是基于这个原因

@zhujian7  确认一下

bbbmj replied:
没有装 go 的话，GOPATH 也是另外设置的好像？

Meoop replied:
或许安装到项目的 bin 目录也是个不错的选择？
```
# install it into ./bin/
curl -sfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh| sh -s vX.Y.Z
```

zhujian7 replied:
基础镜像只安装了 make 和 dind，在这个镜像里执行 make build-linux，make build-linux 已经指定了 `-e GOPATH=/go` ，所以我觉得不需要换成 `go env GOPATH`.

zhujian7 replied:
基础镜像只安装了 make 和 dind，dind 启的 Golang 镜像只有指定了 GOPATH 就行，我觉得不需要换成 `go env GOPATH`.

如果换成了 `go env GOPATH` ，基础镜像除了要装 make , dind 外，还得再安装个 golang，个人感觉没有必要。

Meoop replied:
Makefile 不兼容本地环境么？新安装的 golang 本地是不一定会设置 GOPATH 环境变量的。

bbbmj replied:
使用 `GOPATH`:
- 基础镜像不需要装 go
- 流水线需要额外设置 GOPATH
- 每个研发自己的电脑，需要装 go 并设置 GOPATH

使用 `go env GOPATH`:
- 基础镜像需要装 go
- 流水线不需要额外设置 GOPATH
- 每个研发自己的电脑，只需要装 go，不需要设置 GOPATH

综合看，我觉得设置成 `go env GOPATH` 可能好一点？

ijumps replied:
不用纠结吧，加个判断，存在 go 就 `go env GOPATH`，不存在就按照之前的

zhujian7 replied:
@Meoop 乐乐说的这种方式可以吗？

Meoop replied:
en，和上面写法比较类似吧，`GOPATH ?= $(shell go env GOPATH)`，如果有 GOPATH 这个环境变量，就不会执行 shell 了，也就不检测 go 存不存在，只有不存在 GOPATH 的时候才去调用 go 命令。

zhujian7 commented:
/hold

supereagle commented:
流水线上肯定不存在这个问题，本地开发环境的话，建议手动设置 GOPATH（一次性操作）。因此，没有必要通过 `go env GOPATH` 自动获取，建议 close 这个 PR。@Meoop WDYT?

Meoop commented:
说实话，我并不认为 “本地开发环境的话，建议手动设置 GOPATH（一次性操作）” 是一个好的方案...

如果远程是通过启动一个 Golang 镜像去跑 lint ，那还不如直接用 golangci-lint 镜像呢，还少一次 curl。

lsytj0413 commented:
我们不是准备切换到 go module 了么，到时候 GOPATH 直接基本就没用了；那么这段过渡时间是否值得改这个？
然后基础镜像为啥要装 go 啊......

Meoop commented:
> 我们不是准备切换到 go module 了么，到时候 GOPATH 直接基本就没用了；那么这段过渡时间是否值得改这个？

@lsytj0413  这个改动就是由于系统不设置 GOPATH 这个环境变量导致的，切换到 go module，GOPATH 这个环境变量估计也没有人会去设置了，到时候根据 $GOPATH 环境变量去保存 golangci-lint 二进制就会出现问题（$GOPATH 为空，下载到 /bin 下，普通用户是没有权限的）。

zhujian7 commented:
![image](https://user-images.githubusercontent.com/36154065/64684059-fcaf6400-d4b6-11e9-853a-c1715cadab45.png)

验证完了，对 `build-linux` 没有影响，可以加 `go env GOPATH`
/hold cancel

/cc @Meoop @supereagle @bbbmj

bbbmj commented:
/lgtm

Meoop committed:
fix(*): Use go env GOPATH for install golangci-lint

GOPATH environment variable may be empty, go env GOPATH will work for all cases.

Meoop commented:
/reopen

supereagle commented:
/lgtm

bbbmj commented:
@Meoop please cherry-pick to https://github.com/caicloud/nirvana-template-project

supereagle commented:
> @Meoop please cherry-pick to https://github.com/caicloud/nirvana-template-project

I will update Makefile for this project including my changes.

supereagle commented:
Cherry pick to [caicloud/nirvana-template-project](https://github.com/caicloud/nirvana-template-project) by caicloud/nirvana-template-project#6
